File: Lib/gftools/scripts/unicode_names.py
#!/usr/bin/env python3
#
# Copyright 2017 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""Utility to print unicode character names from a nam file.

Input file should have one codepoint per line in hex (0xXXXX).

"""
from __future__ import print_function
import unicodedata
import sys
import argparse

parser = argparse.ArgumentParser(
    description="Add Unicode character names to a nam file"
)
parser.add_argument("--nam_file", help="Location of nam file")


def main(args=None):
    args = parser.parse_args(args)
    with open(args.nam_file, "r") as f:
        for line in f:
            print(_ReformatLine(line))


def _ReformatLine(line):
    if line.startswith("0x"):
        codepoint = int(line[2:6], 16)  # This'll only work for BMP...
        out = chr(codepoint) + " " + unicodedata.name(chr(codepoint), "")
        return "0x%04X  %s" % (codepoint, out)
    else:
        return line


if __name__ == "__main__":
    main()

File: src/bin/unicode_names.rs
use clap::Parser;

#[derive(Debug, Parser)]
#[command(version, about, long_about = None)]
/// Utility to print unicode character names from a nam file.
///
/// Input file should have one codepoint per line in hex (0xXXXX).
struct Args {
    nam_file: String,
}

fn main() {
    let args = Args::parse();
    let nam_file = std::fs::read_to_string(args.nam_file).expect("Failed to read nam file");
    for line in nam_file.lines() {
        if let Ok(codepoint) = u32::from_str_radix(line.trim_start_matches("0x"), 16) {
            if let Some(name) = char::from_u32(codepoint).and_then(unicode_names2::name) {
                println!("U+{:04X}: {}", codepoint, name);
            } else {
                println!("U+{:04X}: <unknown>", codepoint);
            }
        } else {
            eprintln!("Invalid line in nam file: {}", line);
        }
    }
}
